Hier alle imports gewoon plaatsen denk ik, zouden het ook kunnen doen per deel maar niet dat het in random codeblocks staat

**Statlines**
- Bevolking = https://opendata.cbs.nl/#/CBS/nl/dataset/37230ned/table?dl=A564B 2019-2023 Niet gebruikt
- Inkomen = https://opendata.cbs.nl/#/CBS/nl/dataset/86005NED/table?dl=9552C 2021 Niet gebruikt
- Woningen = https://opendata.cbs.nl/#/CBS/nl/dataset/85035NED/table?dl=B40B6 2021
- Zon = https://opendata.cbs.nl/#/CBS/nl/dataset/85005NED/table?dl=B3FBC 2019-2021
- Inkomen & Leeftijd & Bevolking & Onderwijs & oppervlaktes samen = https://opendata.cbs.nl/#/CBS/nl/dataset/70072ned/table?dl=B40C8 2019-2023 (Nu gebruikt)
- Dichtheid kan je uitrekenen met oppervlakte en woningen
- Missend zijn subsidie en dakorientatie

IAPES (engels) wordt gebruikt voor de DF met Inkomen, leeftijd, Bevolking, onderwijs, oppervlaktes

In [101]:
import pandas as pd
import numpy as np
import geopandas as gpd



In [102]:
# Path to the relevant geofiles
municipality_path = "data/geo/gemeente_2020_v3.shp"

# Path to the relevant files
housing_type_path = "data/woningtype_nieuw.csv"
solar_path = "data/zonnestroom.csv"
IAPES_path = "data/IAPES.csv"


In [103]:
# Reading the shapefile as a GeoDataFrame
geo_df = gpd.read_file(municipality_path)


# Reading the other datasets as DataFrames
housing_type_df = pd.read_csv(housing_type_path, header=0, delimiter=';')
solar_df = pd.read_csv(solar_path, header=0, delimiter=';')
IAPES_df = pd.read_csv(IAPES_path, header=0, delimiter=';')


In [104]:
display(geo_df)

GM_CODE             GM_NAAM  H2O       OAD      STED  BEV_DICHTH  \
0    GM0010            Delfzijl   JA -99999999 -99999999   -99999999   
1    GM0034              Almere   JA -99999999 -99999999   -99999999   
2    GM0050            Zeewolde   JA -99999999 -99999999   -99999999   
3    GM0060             Ameland   JA -99999999 -99999999   -99999999   
4    GM0072           Harlingen   JA -99999999 -99999999   -99999999   
..      ...                 ...  ...       ...       ...         ...   
433  GM1963      Hoeksche Waard  NEE       825         4         325   
434  GM1966        Het Hogeland  NEE       410         5          99   
435  GM1969      Westerkwartier  NEE       475         5         175   
436  GM1970  Noardeast-Fryslân  NEE       463         5         120   
437  GM1978         Molenlanden  NEE       391         5         242   

     AANT_INW  AANT_MAN  AANT_VROUW  P_00_14_JR  ...  AV20PODIUM   AF_MUSEUM  \
0   -99999999 -99999999   -99999999   -99999999  ... -99999999.0 -99999999.0   
1   -99999999 -99999999   -99999999   -99999999  ... -99999999.0 -99999999.0   
2   -99999999 -99999999   -99999999   -99999999  ... -99999999.0 -99999999.0   
3   -99999999 -99999999   -99999999   -99999999  ... -99999999.0 -99999999.0   
4   -99999999 -99999999   -99999999   -99999999  ... -99999999.0 -99999999.0   
..        ...       ...         ...         ...  ...         ...         ...   
433     87401     43211       44190          16  ...         9.8         7.9   
434     47801     24042       23759          15  ...         3.3         5.9   
435     63329     31810       31519          17  ...         5.6         5.2   
436     45228     22768       22460          17  ...         1.2         4.6   
437     43909     22256       21653          18  ...        10.4         7.0   

     AV5_MUSEUM  AV10MUSEUM  AV20MUSEUM  JRSTATCODE  JAAR     Shape_Leng  \
0   -99999999.0 -99999999.0 -99999999.0  2020GM0010  2020  128728.007061   
1   -99999999.0 -99999999.0 -99999999.0  2020GM0034  2020  122665.358634   
2   -99999999.0 -99999999.0 -99999999.0  2020GM0050  2020   83290.773617   
3   -99999999.0 -99999999.0 -99999999.0  2020GM0060  2020  156406.894023   
4   -99999999.0 -99999999.0 -99999999.0  2020GM0072  2020  104274.683036   
..          ...         ...         ...         ...   ...            ...   
433         0.2         1.0        10.4  2020GM1963  2020  147020.366016   
434         0.4         1.7         7.0  2020GM1966  2020  323086.469173   
435         0.8         2.4        12.9  2020GM1969  2020   99030.762281   
436         0.7         1.9         7.5  2020GM1970  2020  238426.959616   
437         0.4         1.8        15.5  2020GM1978  2020   88570.375655   

       Shape_Area                                           geometry  
0    9.129235e+07  MULTIPOLYGON (((268674.309 592382.331, 268667....  
1    1.095623e+08  MULTIPOLYGON (((150213.998 479503.726, 150087....  
2    1.633050e+07  MULTIPOLYGON (((154927.535 474979.29, 154931.2...  
3    2.085914e+08  POLYGON ((196000 610000, 196000 608000, 195250...  
4    3.618323e+08  MULTIPOLYGON (((158392.775 580357.5, 158387.71...  
..            ...                                                ...  
433  2.818623e+08  MULTIPOLYGON (((98666.226 412682.437, 98543 41...  
434  4.878506e+08  MULTIPOLYGON (((215169.098 595155.688, 215169....  
435  3.688457e+08  POLYGON ((211363.51 585818.36, 211364.156 5858...  
436  3.837930e+08  MULTIPOLYGON (((207521.119 595796.903, 207521....  
437  1.915841e+08  POLYGON ((102822.359 432749.868, 102817.412 43...  

[438 rows x 205 columns]

In [105]:
display(housing_type_df)

Regio's                           Woningtype  Perioden  \
0      Nederland                      Totaal woningen      2021   
1      Nederland            Meergezinswoningen totaal      2021   
2      Nederland             Eengezinswoningen totaal      2021   
3      Nederland                   Vrijstaande woning      2021   
4      Nederland                  2-onder-1-kapwoning      2021   
...          ...                                  ...       ...   
2859  Buitenland                   Vrijstaande woning      2021   
2860  Buitenland                  2-onder-1-kapwoning      2021   
2861  Buitenland                           Hoekwoning      2021   
2862  Buitenland                         Tussenwoning      2021   
2863  Buitenland  Onbekend woningtype eengezinswoning      2021   

      Beginstand woningvoorraad (aantal)  
0                              7966331.0  
1                              2870532.0  
2                              5095799.0  
3                              1035389.0  
4                               698022.0  
...                                  ...  
2859                                 NaN  
2860                                 NaN  
2861                                 NaN  
2862                                 NaN  
2863                                 NaN  

[2864 rows x 4 columns]

In het begin werkten de tabellen kut maar met andere download is dit niet echt meer nodig. Moest unstacken

In [106]:
# def fix_columns(df, add_unit=True):
#     # Adjust the headers
#     # Extract row 0 values for column names
#     row_0 = df.iloc[0]

#     # Set new column names
#     new_column_names = [
#         "Perioden" if idx == 0 else
#         "Regio's" if idx == 1 else
#         (f"{df.columns[idx]} ({row_0[idx]})" if add_unit else df.columns[idx])
#         for idx in range(len(df.columns))
#     ]

#     # Update the DataFrame with the new headers
#     df.columns = new_column_names

#     # Drop the first row (row 0)
#     df = solar_df[1:]

#     # Reset index for clarity
#     df.reset_index(drop=True, inplace=True)

#     return df

# housing_type_df = fix_columns(housing_type_df)
# solar_df = fix_columns(solar_df)

In [107]:

# Unstacking based on 'Woningtype'
housing_type_df = housing_type_df.pivot_table(
    index=["Regio's", 'Perioden'], 
    columns='Woningtype', 
    values='Beginstand woningvoorraad (aantal)'
).reset_index()

display(housing_type_df)

Woningtype                   Regio's  Perioden  2-onder-1-kapwoning  \
0           's-Gravenhage (gemeente)      2021               3063.0   
1                   's-Hertogenbosch      2021               3850.0   
2                        Aa en Hunze      2021               2621.0   
3                           Aalsmeer      2021               1111.0   
4                             Aalten      2021               2313.0   
..                               ...       ...                  ...   
348                          Zundert      2021               1742.0   
349                          Zutphen      2021               1802.0   
350                  Zwartewaterland      2021               1428.0   
351                      Zwijndrecht      2021                744.0   
352                           Zwolle      2021               2720.0   

Woningtype  Eengezinswoningen totaal  Hoekwoning  Meergezinswoningen totaal  \
0                            56307.0      9259.0                   209652.0   
1                            44343.0      9400.0                    29081.0   
2                            10262.0       925.0                      945.0   
3                            10389.0      2339.0                     2895.0   
4                             9893.0      1914.0                     2135.0   
..                               ...         ...                        ...   
348                           8420.0      1116.0                     1229.0   
349                          15234.0      3587.0                     7417.0   
350                           7917.0      1766.0                     1096.0   
351                          12479.0      3345.0                     8241.0   
352                          38081.0      8383.0                    21410.0   

Woningtype  Onbekend woningtype eengezinswoning  Totaal woningen  \
0                                          10.0         265959.0   
1                                          16.0          73424.0   
2                                           1.0          11207.0   
3                                           5.0          13284.0   
4                                           4.0          12028.0   
..                                          ...              ...   
348                                         2.0           9649.0   
349                                         7.0          22651.0   
350                                        39.0           9013.0   
351                                        12.0          20720.0   
352                                        14.0          59491.0   

Woningtype  Tussenwoning  Vrijstaande woning  
0                41662.0              2313.0  
1                25311.0              5766.0  
2                 1149.0              5566.0  
3                 4972.0              1962.0  
4                 2587.0              3075.0  
..                   ...                 ...  
348               1679.0              3881.0  
349               8361.0              1477.0  
350               3180.0              1504.0  
351               7718.0               660.0  
352              23804.0              3160.0  

[353 rows x 10 columns]

In [108]:
display(solar_df)

Sector en vermogensklasse                      Regio's  \
0     Alle economische activiteit en woningen                    Nederland   
1     Alle economische activiteit en woningen                    Nederland   
2     Alle economische activiteit en woningen                    Nederland   
3     Alle economische activiteit en woningen                    Nederland   
4     Alle economische activiteit en woningen                  Aa en Hunze   
...                                       ...                          ...   
1443  Alle economische activiteit en woningen                       Zwolle   
1444  Alle economische activiteit en woningen  Gemeenten; niet in te delen   
1445  Alle economische activiteit en woningen  Gemeenten; niet in te delen   
1446  Alle economische activiteit en woningen  Gemeenten; niet in te delen   
1447  Alle economische activiteit en woningen  Gemeenten; niet in te delen   

      Perioden  Installaties (aantal)  \
0         2019              1062265.0   
1         2020              1383888.0   
2         2021              1730285.0   
3         2022              2298859.0   
4         2019                 2922.0   
...        ...                    ...   
1443      2022                19666.0   
1444      2019                  359.0   
1445      2020                  432.0   
1446      2021                  426.0   
1447      2022                  509.0   

      Opgesteld vermogen van zonnepanelen (kWp)  \
0                                     7225864.0   
1                                    11108427.0   
2                                    14822880.0   
3                                    19536096.0   
4                                       20968.0   
...                                         ...   
1443                                   172154.0   
1444                                     1303.0   
1445                                     1823.0   
1446                                    14389.0   
1447                                    17208.0   

      Opgesteld vermogen omvormers (kW) Productie van zonnestroom (mln kWh)  
0                                   NaN                                5399  
1                                   NaN                                8567  
2                                   NaN                               11304  
3                            17451365.0                               16659  
4                                   NaN                                   .  
...                                 ...                                 ...  
1443                           150845.0                                   .  
1444                                NaN                                   1  
1445                                NaN                                   1  
1446                                NaN                                  12  
1447                            14768.0                                  16  

[1448 rows x 7 columns]

In [109]:
display(IAPES_df)

Perioden    Regio's  \
0         2019  Nederland   
1         2020  Nederland   
2         2021  Nederland   
3         2022  Nederland   
4         2023  Nederland   
...        ...        ...   
3635      2019     Zwolle   
3636      2020     Zwolle   
3637      2021     Zwolle   
3638      2022     Zwolle   
3639      2023     Zwolle   

      Bevolking/Bevolkingssamenstelling op 1 januari/Totale bevolking (aantal)  \
0                                            17282163.0                          
1                                            17407585.0                          
2                                            17475415.0                          
3                                            17590672.0                          
4                                            17811291.0                          
...                                                 ...                          
3635                                           127497.0                          
3636                                           128840.0                          
3637                                           129840.0                          
3638                                           130668.0                          
3639                                           132411.0                          

      Bevolking/Bevolkingssamenstelling op 1 januari/Leeftijd/Leeftijdsgroepen, relatief/Jonger dan 5 jaar (%)  \
0                                                   5.0                                                          
1                                                   4.9                                                          
2                                                   4.9                                                          
3                                                   4.9                                                          
4                                                   4.9                                                          
...                                                 ...                                                          
3635                                                5.6                                                          
3636                                                5.5                                                          
3637                                                5.4                                                          
3638                                                5.4                                                          
3639                                                5.3                                                          

      Bevolking/Bevolkingssamenstelling op 1 januari/Leeftijd/Leeftijdsgroepen, relatief/5 tot 10 jaar (%)  \
0                                                   5.3                                                      
1                                                   5.2                                                      
2                                                   5.1                                                      
3                                                   5.1                                                      
4                                                   5.0                                                      
...                                                 ...                                                      
3635                                                6.1                                                      
3636                                                6.0                                                      
3637                                                5.8                                                      
3638                                                5.6                                                      
3639                                                5.4                                      

In [110]:
# Filtering for 2021
housing_type_df_2021 = housing_type_df[housing_type_df['Perioden'] == 2021]
solar_df_2021 = solar_df[solar_df['Perioden'] == 2021]
IAPES_df_2021 = IAPES_df[IAPES_df['Perioden'] == 2021]

# Merging on 'Regio's'
merged_df = housing_type_df_2021.merge(
    solar_df_2021, on="Regio's", suffixes=('_housing', '_solar')
).merge(
    IAPES_df_2021, on="Regio's", suffixes=('', '_IAPES')
)

Als je merged verlies je 1448/4=362 362-353 (kijk naar rij aantallen)=9 gemeenten, waaraan weet ik nog niet. Huisdata lijkt iig het minste aantal gemeenten te hebben, wel staan in veel oude cijfers gemeenten nog wel maar zonder data. Na een kleine google kom ik op 352 gemeenten nodig in 2021, en aangezien Nederland er ook bij zit klopt 353 precies (bron https://allecijfers.nl/definities/aantal-gemeenten-en-overzicht-gemeentelijkeherindelingen/)

In [111]:
display(merged_df)

Regio's  Perioden_housing  2-onder-1-kapwoning  \
0    's-Gravenhage (gemeente)              2021               3063.0   
1            's-Hertogenbosch              2021               3850.0   
2                 Aa en Hunze              2021               2621.0   
3                    Aalsmeer              2021               1111.0   
4                      Aalten              2021               2313.0   
..                        ...               ...                  ...   
348                   Zundert              2021               1742.0   
349                   Zutphen              2021               1802.0   
350           Zwartewaterland              2021               1428.0   
351               Zwijndrecht              2021                744.0   
352                    Zwolle              2021               2720.0   

     Eengezinswoningen totaal  Hoekwoning  Meergezinswoningen totaal  \
0                     56307.0      9259.0                   209652.0   
1                     44343.0      9400.0                    29081.0   
2                     10262.0       925.0                      945.0   
3                     10389.0      2339.0                     2895.0   
4                      9893.0      1914.0                     2135.0   
..                        ...         ...                        ...   
348                    8420.0      1116.0                     1229.0   
349                   15234.0      3587.0                     7417.0   
350                    7917.0      1766.0                     1096.0   
351                   12479.0      3345.0                     8241.0   
352                   38081.0      8383.0                    21410.0   

     Onbekend woningtype eengezinswoning  Totaal woningen  Tussenwoning  \
0                                   10.0         265959.0       41662.0   
1                                   16.0          73424.0       25311.0   
2                                    1.0          11207.0        1149.0   
3                                    5.0          13284.0        4972.0   
4                                    4.0          12028.0        2587.0   
..                                   ...              ...           ...   
348                                  2.0           9649.0        1679.0   
349                                  7.0          22651.0        8361.0   
350                                 39.0           9013.0        3180.0   
351                                 12.0          20720.0        7718.0   
352                                 14.0          59491.0       23804.0   

     Vrijstaande woning  ...  \
0                2313.0  ...   
1                5766.0  ...   
2                5566.0  ...   
3                1962.0  ...   
4                3075.0  ...   
..                  ...  ...   
348              3881.0  ...   
349              1477.0  ...   
350              1504.0  ...   
351               660.0  ...   
352              3160.0  ...   

    Onderwijs/Naar woongemeente/Gediplomeerden/Voortgezet onderwijs (aantal)  \
0                                               4530.0                         
1                                               1401.0                         
2                                                245.0                         
3                                                405.0                         
4                                                291.0                         
..                                                 ...                         
348                                              172.0                         
349                                              491.0                         
350                                              315.0                         
351                                              463.0                         
352                                             1447.0                         

     Onderwijs/Naar woongemeente/Gediplomeerd